src/mpid/ch4/shm/xpmem/xpmem_recv.h

```diff
@@ -9,71 +9,281 @@
+    MPIR_ERR_CHECK(mpi_errno);
+
+    MPIDI_Datatype_check_size_lb(MPIDIG_REQUEST(sreq, datatype), MPIDIG_REQUEST(sreq, count),
+                                 data_sz, dt_true_lb);
+
+    src_buf = (void *) ((char *) MPIDIG_REQUEST(sreq, buffer) + dt_true_lb);
+    counter_ptr = &MPIDI_XPMEM_global.coop_counter[dest_lrank * MPIDI_XPMEM_global.num_local +
+                                                   MPIDI_XPMEM_global.local_rank];
+
+    /* sender and receiver datatypes are both continuous, perform cooperative copy. */
+    mpi_errno = MPIDI_XPMEM_do_lmt_coop_copy(src_buf, data_sz, dest_buf, counter_ptr,
+                                             MPIDI_SHM_XPMEM_SEND_LMT_RECV_FIN, req_ptr,
+                                             sreq, MPIDI_XPMEM_ctrl_send_lmt_send_fin_cb);
+    MPIR_ERR_CHECK(mpi_errno);
+
+    mpi_errno = MPIDI_XPMEM_seg_deregist(seg_ptr);
+    MPIR_ERR_CHECK(mpi_errno);
+
+  fn_exit:
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_XPMEM_HANDLE_LMT_CTS_RECV);
+    return mpi_errno;
+  fn_fail:
+    goto fn_exit;
+}
+
+MPL_STATIC_INLINE_PREFIX int MPIDI_XPMEM_handle_lmt_coop_recv(uint64_t src_offset,
+                                                              size_t src_data_sz, uint64_t sreq_ptr,
+                                                              int src_lrank, MPIR_Comm * comm,
+                                                              MPIR_Request * rreq)
+{
+    int mpi_errno = MPI_SUCCESS;
+    int dt_true_lb;
+    MPIDI_XPMEM_seg_t *seg_ptr = NULL;
+    void *src_buf = NULL;
+    void *dest_buf = NULL;
     size_t data_sz, recv_data_sz;
-    MPIDI_SHM_ctrl_hdr_t ack_ctrl_hdr;
-    MPIDI_SHM_ctrl_xpmem_send_lmt_ack_t *slmt_ack_hdr = &ack_ctrl_hdr.xpmem_slmt_ack;
 
-    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_XPMEM_HANDLE_LMT_RECV);
-    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_XPMEM_HANDLE_LMT_RECV);
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_XPMEM_HANDLE_LMT_COOP_RECV);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_XPMEM_HANDLE_LMT_COOP_RECV);
 
-    mpi_errno = MPIDI_XPMEM_seg_regist(src_lrank, src_data_sz, (void *) src_offset,
-                                       &seg_ptr, &attached_sbuf);
+    MPIDI_Datatype_check_size_lb(MPIDIG_REQUEST(rreq, datatype), MPIDIG_REQUEST(rreq, count),
+                                 data_sz, dt_true_lb);
+
+    mpi_errno = MPIDI_XPMEM_lmt_cts_isend(MPIDIG_REQUEST(rreq, buffer),
+                                          MPIDIG_REQUEST(rreq, count),
+                                          MPIDIG_REQUEST(rreq, datatype),
+                                          MPIDIG_REQUEST(rreq, rank), comm, sreq_ptr, rreq);
     MPIR_ERR_CHECK(mpi_errno);
 
-    MPIDI_Datatype_check_size(MPIDIG_REQUEST(rreq, datatype), MPIDIG_REQUEST(rreq, count), data_sz);
+    /* Data truncation checking */
+    recv_data_sz = MPL_MIN(src_data_sz, data_sz);
     if (src_data_sz > data_sz)
         rreq->status.MPI_ERROR = MPI_ERR_TRUNCATE;
 
```

Comment:
When `src_data_sz == 0`, maybe we should skip `CTS` and directly reply `FIN` to sender.

Reply:
If we are using XPMEM, the message size must be larger than `MPIR_CVAR_CH4_XPMEM_LMT_MSG_SIZE`. I think it's ok to think src_data_sz > 0 by default

Reply:
It is a little tricky. We currently set `-1` to disable XPMEM, and the smallest data size can be set is zero. But we compare it as `greater` rather than `greater-equal`, thus the data size can never be zero. I'd suggest assume `src_data_sz==0` as a valid case on the receiver side, in case someone changes the comparison on the sender side in future.

Reply:
I am confused. If XPMEM is disabled, program will never call any XPMEM routine right? How does this cause a bug?

Reply:
What I meant is that someone may change the sender side check in future but does not notice that the receiver side does not handle the zero-byte message case. For example, if I change the sender side code as below and set `MPIR_CVAR_CH4_XPMEM_LMT_MSG_SIZE=0`
```
    if (MPIR_CVAR_CH4_XPMEM_LMT_MSG_SIZE > -1 &&
        dt_contig && data_sz >= MPIR_CVAR_CH4_XPMEM_LMT_MSG_SIZE) {
        mpi_errno = MPIDI_XPMEM_lmt_rts_isend(buf, count, datatype, rank, tag, comm,
                                              context_offset, addr, request);
        MPIR_ERR_CHECK(mpi_errno);
    }
```
Note that, zero-byte message is valid in MPI, it is often considered as a peer-synchronization in programs.

Reply:
Emm, I think the current code might work correctly with zero-byte message but is just not optimized (i.e., skip unnecessary CTS and counter atomic fetch). Can you please test?

It is better if we can optimize it in the PR.